// @contentstack/management/dist/es5/stack/variantGroup/variants/index.js
"use strict";

var _interopRequireDefault3 = require("@babel/runtime/helpers/interopRequireDefault");
var _interopRequireDefault2 = _interopRequireDefault3(require("@babel/runtime/helpers/interopRequireDefault"));
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _defineProperty2 = require("@babel/runtime/helpers/defineProperty");
var _defineProperty3 = (0, _interopRequireDefault2["default"])(_defineProperty2);
var _asyncToGenerator2 = require("@babel/runtime/helpers/asyncToGenerator");
var _asyncToGenerator3 = (0, _interopRequireDefault2["default"])(_asyncToGenerator2);
exports.Variants = Variants;
exports.VariantsCollection = VariantsCollection;
var _regenerator = require("@babel/runtime/regenerator");
var _regenerator2 = (0, _interopRequireDefault2["default"])(_regenerator);
var _cloneDeep = require("lodash/cloneDeep");
var _cloneDeep2 = (0, _interopRequireDefault2["default"])(_cloneDeep);
var _entity = require("../../../entity");
var _contentstackError = require("../../../core/contentstackError");
var _contentstackError2 = (0, _interopRequireDefault2["default"])(_contentstackError);
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty3["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * Contentstack has a sophisticated multilingual capability. It allows you to create and publish entries in any language. This feature allows you to set up multilingual websites and cater to a wide variety of audience by serving content in their local language(s). Read more about <a href='https://www.contentstack.com/docs/developers/multi-language-content'>VariantGroups</a>.
 * @namespace Variants
 */

function Variants(http) {
  var _this = this;
  var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  Object.assign(this, (0, _cloneDeep2["default"])(data.variants));
  this.stackHeaders = data.stackHeaders;
  this.variant_group_uid = data.variant_group_uid;
  this.urlPath = "/variant_groups/".concat(this.variant_group_uid, "/variants");
  if (data.variants) {
    this.urlPath += "/".concat(this.uid);
    /**
       * @description The Update Variants call lets you update the name and description of an existing Variants.
       * @memberof Variants
       * @func update
       * @returns {Promise<Variants.Variants>} Promise for Variants instance
       * @example
       * import * as contentstack from '@contentstack/management'
       * const client = contentstack.client()
       * const data = { "name": "Update name" }
       * client.stack({ api_key: 'api_key'}).VariantGroup('variant_group_uid').variants('variant_uid').update(data)
       * .then((variants) => console.log(variants))
       *
       */
    this.update = /*#__PURE__*/function () {
      var _ref = (0, _asyncToGenerator3["default"])(/*#__PURE__*/_regenerator2["default"].mark(function _callee(data) {
        var response, _t;
        return _regenerator2["default"].wrap(function (_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.prev = 0;
              _context.next = 1;
              return http.put(_this.urlPath, data, {
                headers: _objectSpread({}, (0, _cloneDeep2["default"])(_this.stackHeaders))
              });
            case 1:
              response = _context.sent;
              if (!response.data) {
                _context.next = 2;
                break;
              }
              return _context.abrupt("return", response.data);
            case 2:
              return _context.abrupt("return", (0, _contentstackError2["default"])(response));
            case 3:
              _context.next = 5;
              break;
            case 4:
              _context.prev = 4;
              _t = _context["catch"](0);
              return _context.abrupt("return", (0, _contentstackError2["default"])(_t));
            case 5:
            case "end":
              return _context.stop();
          }
        }, _callee, null, [[0, 4]]);
      }));
      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }();

    /**
       * @description The fetch Variants call fetches Variants details.
       * @memberof Variants
       * @func fetch
       * @returns {Promise<Variants.Variants>} Promise for Variants instance
       * @param {Int} version Enter the unique ID of the content type of which you want to retrieve the details. The UID is generated based on the title of the content type. The unique ID of a content type is unique across a stack.
       * @example
       * import * as contentstack from '@contentstack/management'
       * const client = contentstack.client()
       *
       * client.stack({ api_key: 'api_key'}).VariantGroup('variant_group_uid').variants('variant_uid').fetch()
       * .then((variants) => console.log(variants))
       *
       */
    this.fetch = /*#__PURE__*/(0, _asyncToGenerator3["default"])(/*#__PURE__*/_regenerator2["default"].mark(function _callee2() {
      var param,
        headers,
        response,
        _args2 = arguments,
        _t2;
      return _regenerator2["default"].wrap(function (_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            param = _args2.length > 0 && _args2[0] !== undefined ? _args2[0] : {};
            _context2.prev = 1;
            headers = {
              headers: _objectSpread({}, (0, _cloneDeep2["default"])(_this.stackHeaders)),
              params: _objectSpread({}, (0, _cloneDeep2["default"])(param))
            } || {};
            _context2.next = 2;
            return http.get(_this.urlPath, headers);
          case 2:
            response = _context2.sent;
            if (!response.data) {
              _context2.next = 3;
              break;
            }
            return _context2.abrupt("return", response.data);
          case 3:
            throw (0, _contentstackError2["default"])(response);
          case 4:
            _context2.next = 6;
            break;
          case 5:
            _context2.prev = 5;
            _t2 = _context2["catch"](1);
            (0, _contentstackError2["default"])(_t2);
          case 6:
          case "end":
            return _context2.stop();
        }
      }, _callee2, null, [[1, 5]]);
    }));

    /**
       * @description The Delete Variant call is used to delete an existing Variant permanently from your Stack.
       * @memberof VariantGroup
       * @func delete
       * @returns {Object} Response Object.
       * @example
       * import * as contentstack from '@contentstack/management'
       * const client = contentstack.client()
       *
       * client.stack({ api_key: 'api_key'}).VariantGroup('variant_group_uid').variants('variant_uid').delete()
       * .then((response) => console.log(response.notice))
       */
    this["delete"] = (0, _entity.deleteEntity)(http);
  } else {
    /**
     * @description The Create a variant group call creates a new variant group in a particular stack of your Contentstack account.
     * @memberof Variants
     * @func create
     * @returns {Promise<Variants.Variants>} Promise for Variants instance
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
     * const data = {
     *      "uid": "iphone_color_white", // optional
     *      "name": "White",
     *      "personalize_metadata": {   // optional sent from personalize while creating variant
     *          "experience_uid": "exp1",
     *          "experience_short_uid": "expShortUid1",
     *          "project_uid": "project_uid1",
     *          "variant_short_uid": "variantShort_uid1"
     *      },
     *      }
     * client.stack().VariantGroup('variant_group_uid').variants().create({ data } )
     * .then((variants) => console.log(variants))
     */
    this.create = /*#__PURE__*/function () {
      var _ref3 = (0, _asyncToGenerator3["default"])(/*#__PURE__*/_regenerator2["default"].mark(function _callee3(data) {
        var response, _t3;
        return _regenerator2["default"].wrap(function (_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              _context3.prev = 0;
              _context3.next = 1;
              return http.post("".concat(_this.urlPath), data, {
                headers: _objectSpread({}, (0, _cloneDeep2["default"])(_this.stackHeaders))
              });
            case 1:
              response = _context3.sent;
              if (!response.data) {
                _context3.next = 2;
                break;
              }
              return _context3.abrupt("return", response.data);
            case 2:
              return _context3.abrupt("return", (0, _contentstackError2["default"])(response));
            case 3:
              _context3.next = 5;
              break;
            case 4:
              _context3.prev = 4;
              _t3 = _context3["catch"](0);
              return _context3.abrupt("return", (0, _contentstackError2["default"])(_t3));
            case 5:
            case "end":
              return _context3.stop();
          }
        }, _callee3, null, [[0, 4]]);
      }));
      return function (_x2) {
        return _ref3.apply(this, arguments);
      };
    }();

    /**
     * @description The Query on Variant Groups will allow to fetch details of all or specific Variant Groups
     * @memberof Variants
     * @func query
     * @param {Boolean} include_count Set this to 'true' to include in response the total count of content types available in your stack.
     * @returns {Array<VariantGroup>} Array of ContentTyoe.
     *
     * @example
     * import * as contentstack from '@contentstack/management'
     * const client = contentstack.client()
      * client.stack(api_key).VariantGroup('variant_group_uid').variants().query({ query: { name: 'white' } }).find()
     * .then((variant_groups) => console.log(variant_groups))
     */
    this.query = (0, _entity.query)({
      http: http,
      wrapperCollection: VariantsCollection
    });
  }
  return this;
}
function VariantsCollection(http, data) {
  var obj = (0, _cloneDeep2["default"])(data.variants) || [];
  var variantsCollection = obj.map(function (userdata) {
    return new Variants(http, {
      variants: userdata,
      variant_group_uid: data.uid,
      stackHeaders: data.stackHeaders
    });
  });
  return variantsCollection;
}
